### ISSUE TITLE
NoWhitespaceAfter: add ability to enforce no whitespaces in array brackets

### ISSUE THREAD

**Issue by eric-milles** [checkstyle/checkstyle#3330]
I'm not sure exactly how to classify this one.  I believe this is a request for new functionality similar to ParenPad or GenericWhitespace.  I'd like to be able to check that array defs and uses do not contain extra whitespaces.

I can use this to check for space between the ident and brackets.
```xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="NoWhitespaceAfter">
            <message key="ws.followed" value="Whitespace after ''{0}'' should be removed" />
            <property name="tokens" value="ARRAY_DECLARATOR,INDEX_OP" />
            <property name="allowLineBreaks" value="false" />
        </module>
    </module>
</module>
```

However, to check for some of these things, I am not finding the necessary options in NoWhitespaceAfter and NoWhitespaceBefore.  Looking at the AST, I don't think there is a good token type to request be added, since INDEX_OP is used for the space between the ident and left bracket.

```java
class ArraySubscript
{
    {
        String[ ] array = {"one", "two"}; // VIOLATION -- want no space after opening '[' in array declarator

        String s = array[ 0 ]; // 2 VIOLATIONS -- want no space after opening '[' and no space before closing ']' in index operation

        array[ 0 ] = "value"; // 2 VIOLATIONS -- want no space after opening '[' and no space before closing ']' in index operation
    }
}
```

This doesn't look like it would be easy to find with regexes either, since we have so many string literals that contain square brackets followed by spaces.

**Comment by romani** [comment]
Your case
```
$ java -jar checkstyle-6.17-all.jar -t ArraySubscript.java 
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> ArraySubscript [1:6]
`--OBJBLOCK -> OBJBLOCK [1:11]
    |--LCURLY -> { [1:11]
    |--INSTANCE_INIT -> INSTANCE_INIT [2:1]
    |   `--SLIST -> { [2:1]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [3:10]
    |       |   |--MODIFIERS -> MODIFIERS [3:10]
    |       |   |--TYPE -> TYPE [3:10]
    |       |   |   `--ARRAY_DECLARATOR -> [ [3:10]
    |       |   |       |--IDENT -> String [3:4]
    |       |   |       `--RBRACK -> ] [3:11]
    |       |   |--IDENT -> array [3:13]
    |       |   `--ASSIGN -> = [3:19]
    |       |       `--ARRAY_INIT -> { [3:21]
    |       |           |--EXPR -> EXPR [3:22]
    |       |           |   `--STRING_LITERAL -> "one" [3:22]
    |       |           |--COMMA -> , [3:27]
    |       |           |--EXPR -> EXPR [3:29]
    |       |           |   `--STRING_LITERAL -> "two" [3:29]
    |       |           `--RCURLY -> } [3:34]
    |       |--SEMI -> ; [3:35]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [5:4]
    |       |   |--MODIFIERS -> MODIFIERS [5:4]
    |       |   |--TYPE -> TYPE [5:4]
    |       |   |   `--IDENT -> String [5:4]
    |       |   |--IDENT -> s [5:11]
    |       |   `--ASSIGN -> = [5:13]
    |       |       `--EXPR -> EXPR [5:20]
    |       |           `--INDEX_OP -> [ [5:20]
    |       |               |--IDENT -> array [5:15]
    |       |               |--EXPR -> EXPR [5:22]
    |       |               |   `--NUM_INT -> 0 [5:22]
    |       |               `--RBRACK -> ] [5:24]
    |       |--SEMI -> ; [5:25]
    |       |--EXPR -> EXPR [7:13]
    |       |   `--ASSIGN -> = [7:13]
    |       |       |--INDEX_OP -> [ [7:9]
    |       |       |   |--IDENT -> array [7:4]
    |       |       |   |--EXPR -> EXPR [7:10]
    |       |       |   |   `--NUM_INT -> 0 [7:10]
    |       |       |   `--RBRACK -> ] [7:11]
    |       |       `--STRING_LITERAL -> "value" [7:15]
    |       |--SEMI -> ; [7:22]
    |       `--RCURLY -> } [8:3]
    `--RCURLY -> } [9:0]

```

More exact places of AST:
```
`--ARRAY_DECLARATOR -> [ [3:10]
    |--IDENT -> String [3:4]
    `--RBRACK -> ] [3:11]
```

```
`--INDEX_OP -> [ [5:20]
    |--IDENT -> array [5:15]
    |--EXPR -> EXPR [5:22]
    |   `--NUM_INT -> 0 [5:22]
    `--RBRACK -> ] [5:24]
```

As you see you can enforce no white space after ARRAY_DECLARATOR and INDEX_OP by the NoWhitespaceAfter .
But  NoWhitespaceBefore should be launched on RBRACK, I am not sure that it work without false-positives, you can try.

Summary:
To archive behavior like you want we need to create special check for this. Just an idea ... NoWhitespaceIn that will check no spaces in token (so it should be used only for non-terminal tokens)

I am closing issue, if smb need such Check please open new issue, for new Check creation.

**Comment by eric-milles** [comment]
As with 3326,  I understand from looking at the AST myself that there is no specific token to work with. That is why I opened this issue. What I am after is a specific token or something to set on the check to recognize the RBRACK that is part of INDEX_OP.

I'm not sure I understand the policy of closing issues without any resolution besides "not gonna happen".  Having the priority lowered makes sense, but closing invalidates all the work I put in to explain the situation.  And now I'm no closer to a solution than if I had done nothing.

**Comment by romani** [comment]
> What I am after is a specific token or something to set on the check to recognize the RBRACK that is part of INDEX_OP

new token is not possible to insert as it will damage all other related Checks..

> I'm not sure I understand the policy of closing issues without any resolution besides "not gonna happen". 

as I mentioned above, if you need this Check you need to create new issue on this. You can make reference to this issue if you want keep track of original requests. 
I am not creating the issue myself only due to the fact that I am testing your attitude to this. In most cases users drop idea and disappear.

> And now I'm no closer to a solution than if I had done nothing.

If you need to be closer to solution - please start implementation. I will help and answer your questions.
BTW: http://checkstyle.sourceforge.net/contributing.html#Introduction , please read second attention paragraph.

**Comment by eric-milles** [comment]
It looks as though the array init. braces issues are moving forward separately (3424 and 3325).  So that leaves whitespace in array init. and in array lookups (index op).  Would you see those as handled by separate checks (like ArrayInitPad and ArrayIndexPad) or as one big check similar to GenericWhitespace?  There is a bit of overlap between init (`new String[2]`) and index (`array[2]`).

**Comment by romani** [comment]
>  Would you see those as handled by separate checks (like ArrayInitPad and ArrayIndexPad) or as one big check similar to GenericWhitespace?

it is better to keep Checks separate and be more specific. This let user to change setting to very granular level. On shallow review I like your ideas for new Checks.
Big Checks are problem to support and for user to do what they need. User request for style is unpredictable sometimes.

For now there is moratorium for new Checks in main library, all of them should go to https://github.com/sevntu-checkstyle/sevntu.checkstyle/ project to pass testing of huge amount of sources. Reason - https://groups.google.com/forum/#!topic/checkstyle-devel/JbhfiiSqKsQ
